src/sentry/auth/staff.py

```diff
@@ -0,0 +1,319 @@
+                extra={
+                    "ip_address": request.META["REMOTE_ADDR"],
+                    "user_id": request.user.id,
+                    "expected_user_id": data["uid"],
+                },
+            )
+            return
+
+        if current_datetime is None:
+            current_datetime = timezone.now()
+
+        try:
+            data["idl"] = datetime.utcfromtimestamp(float(data["idl"])).replace(tzinfo=timezone.utc)
+        except (TypeError, ValueError):
+            logger.warning(
+                "staff.invalid-idle-expiration",
+                extra={"ip_address": request.META["REMOTE_ADDR"], "user_id": request.user.id},
+                exc_info=True,
+            )
+            return
+
+        if data["idl"] < current_datetime:
+            logger.info(
+                "staff.session-expired",
+                extra={"ip_address": request.META["REMOTE_ADDR"], "user_id": request.user.id},
+            )
+            return
+
+        try:
+            data["exp"] = datetime.utcfromtimestamp(float(data["exp"])).replace(tzinfo=timezone.utc)
+        except (TypeError, ValueError):
+            logger.warning(
+                "staff.invalid-expiration",
+                extra={"ip_address": request.META["REMOTE_ADDR"], "user_id": request.user.id},
+                exc_info=True,
+            )
+            return
+
+        if data["exp"] < current_datetime:
+            logger.info(
+                "staff.session-expired",
+                extra={"ip_address": request.META["REMOTE_ADDR"], "user_id": request.user.id},
+            )
+            return
+
+        return data
+
+    def _populate(self, current_datetime=None):
+        if current_datetime is None:
+            current_datetime = timezone.now()
+
+        request = self.request
+        user = getattr(request, "user", None)
+        if not hasattr(request, "session"):
+            data = None
+        elif not (user and user.is_staff):
+            data = None
+        else:
+            data = self.get_session_data(current_datetime=current_datetime)
+
+        if not data:
+            self._set_logged_out()
+        else:
+            self._set_logged_in(expires=data["exp"], token=data["tok"], user=user)
+
+            if not self.is_active:
+                if self._inactive_reason:
+                    logger.warning(
+                        "staff.%s",
+                        self._inactive_reason,
+                        extra={
+                            "ip_address": request.META["REMOTE_ADDR"],
+                            "user_id": request.user.id,
+                        },
+                    )
+                else:
+                    logger.warning(
+                        "staff.inactive-unknown-reason",
+                        extra={
+                            "ip_address": request.META["REMOTE_ADDR"],
+                            "user_id": request.user.id,
+                        },
+                    )
```

Comment:
Do we need to do this logic in this function at this level? Right now, `is_activate` property needs `_populate` to finish first so that `_is_active` gets added to the class, now `_populate` needs `is_active` to function correctly. It's creating this cyclic dependency effect, and it would be best to pull that out.